File: ios-native/GatherLists/GatherLists/Utils/SortPipeline.swift
import Foundation

/// Valid sort level options for the sort pipeline.
enum SortLevel: String, Codable, CaseIterable {
    case store
    case category
    case name
    case date
    case price
    case rsvp
}

/// Sort pipeline engine for shopping list items.
/// Provides flexible, composable sorting and grouping of items based on a configurable sort pipeline.
enum SortPipeline {
    /// Maximum levels for grouping (Group By section).
    static let maxGroupLevels = 2
    
    /// Maximum levels for sort-only (Sort By section).
    static let maxSortOnlyLevels = 2
    
    /// Maximum total sort levels.
    static let maxSortLevels = maxGroupLevels + maxSortOnlyLevels
    
    /// System default sort configuration.
    static let systemDefault: [SortLevel] = [.store, .category, .name]
    
    /// Returns the default sort config for a given list type.
    static func getDefaultConfig(for listType: String) -> [SortLevel] {
        let config = ListTypes.getConfig(listType)
        return config.defaultSort.compactMap { SortLevel(rawValue: $0) }
    }
    
    /// Levels that create groups (vs just sorting within existing groups).
    static let groupingLevels: Set<SortLevel> = [.store, .category, .rsvp]
    
    /// Check if a level is a grouping level.
    static func isGroupingLevel(_ level: SortLevel) -> Bool {
        groupingLevels.contains(level)
    }
    
    /// Partition a sort config into group levels and sort-only levels.
    static func partitionConfig(_ config: [SortLevel]) -> (groupLevels: [SortLevel], sortOnlyLevels: [SortLevel]) {
        let groupLevels = config.filter { groupingLevels.contains($0) }
        let sortOnlyLevels = config.filter { !groupingLevels.contains($0) }
        return (groupLevels, sortOnlyLevels)
    }
    
    /// Combine group levels and sort-only levels into a single config array.
    static func combineConfig(groupLevels: [SortLevel], sortOnlyLevels: [SortLevel]) -> [SortLevel] {
        return groupLevels + sortOnlyLevels
    }
    
    /// A group of items created by store or category grouping.
    struct SortGroup {
        let key: String
        let label: String
        let color: String?
        let type: SortLevel
        var items: [Item]?
        var subGroups: [SortGroup]?
    }
    
    /// Result of applying the sort pipeline.
    struct PipelineResult {
        var groups: [SortGroup]
        var ungrouped: [Item]
        var items: [Item]?
    }
    
    // MARK: - Validation
    
    /// Validates a sort configuration.
    static func isValid(_ config: [SortLevel]) -> Bool {
        guard !config.isEmpty, config.count <= maxSortLevels else { return false }
        var seen = Set<SortLevel>()
        for level in config {
            if seen.contains(level) { return false }
            seen.insert(level)
        }
        return true
    }
    
    /// Normalizes any input into a valid sort configuration.
    static func normalize(_ config: [SortLevel]?) -> [SortLevel] {
        guard let config = config else { return systemDefault }
        
        var seen = Set<SortLevel>()
        var normalized: [SortLevel] = []
        
        for level in config {
            guard !seen.contains(level) else { continue }
            seen.insert(level)
            normalized.append(level)
            if normalized.count == maxSortLevels { break }
        }
        
        if normalized.isEmpty { return systemDefault }
        
        // Safety: ensure grouping levels come before sort-only levels
        let groups = normalized.filter { Self.groupingLevels.contains($0) }
        let sorts = normalized.filter { !Self.groupingLevels.contains($0) }
        return groups + sorts
    }
    
    /// Normalizes from string array (for Supabase JSON decoding).
    static func normalize(strings: [String]?) -> [SortLevel] {
        guard let strings = strings else { return systemDefault }
        
        var seen = Set<SortLevel>()
        var normalized: [SortLevel] = []
        
        for str in strings {
            guard let level = SortLevel(rawValue: str), !seen.contains(level) else { continue }
            seen.insert(level)
            normalized.append(level)
            if normalized.count == maxSortLevels { break }
        }
        
        if normalized.isEmpty { return systemDefault }
        
        // Safety: ensure grouping levels come before sort-only levels
        let groups = normalized.filter { Self.groupingLevels.contains($0) }
        let sorts = normalized.filter { !Self.groupingLevels.contains($0) }
        return groups + sorts
    }
    
    // MARK: - Core Function
    
    /// Applies the sort pipeline to items, returning a nested group structure.
    /// - Parameters:
    ///   - items: Array of items to sort/group.
    ///   - config: Array of 1-4 sort levels.
    ///   - stores: Array of store objects for store grouping.
    ///   - listCategories: Categories for the current list (list.categories → user defaults → system defaults).
    ///   - listType: The list type (for type-specific category lookup).
    /// - Returns: Nested structure with groups, subGroups, and items.
    static func apply(items: [Item], config: [SortLevel], stores: [Store], listCategories: [CategoryDef], listType: String = "grocery") -> PipelineResult {
        let normalizedConfig = normalize(config)
        let safeStores = stores
        
        guard !items.isEmpty else {
            return PipelineResult(groups: [], ungrouped: [], items: [])
        }
        
        let hasGroupingLevel = normalizedConfig.contains { groupingLevels.contains($0) }
        
        if !hasGroupingLevel {
            var sorted = items
            for level in normalizedConfig {
                sorted = applySortLevel(sorted, level: level)
            }
            return PipelineResult(groups: [], ungrouped: [], items: sorted)
        }
        
        let result = applyRemainingLevels(
            items: items,
            remainingLevels: normalizedConfig,
            stores: safeStores,
            listCategories: listCategories,
            listType: listType
        )
        
        return PipelineResult(
            groups: result.groups ?? [],
            ungrouped: result.ungrouped ?? []
        )
    }
    
    // MARK: - Private Helpers
    
    private struct IntermediateResult {
        var groups: [SortGroup]?
        var items: [Item]?
        var ungrouped: [Item]?
    }
    
    private static func applySortLevel(_ items: [Item], level: SortLevel) -> [Item] {
        switch level {
        case .name:
            return sortByName(items)
        case .date:
            return sortByDate(items)
        case .price:
            return sortByPrice(items)
        default:
            return items
        }
    }
    
    private static func sortByName(_ items: [Item]) -> [Item] {
        items.sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }
    
    private static func sortByDate(_ items: [Item]) -> [Item] {
        items.sorted { $0.addedAt > $1.addedAt }
    }
    
    private static func sortByPrice(_ items: [Item]) -> [Item] {
        items.sorted { a, b in
            let priceA = a.price
            let priceB = b.price
            if priceA == nil && priceB == nil { return false }
            if priceA == nil { return false }
            if priceB == nil { return true }
            return priceA! < priceB!
        }
    }
    
    private static func applyRemainingLevels(
        items: [Item],
        remainingLevels: [SortLevel],
        stores: [Store],
        listCategories: [CategoryDef],
        listType: String
    ) -> IntermediateResult {
        guard !remainingLevels.isEmpty else {
            return IntermediateResult(items: items)
        }
        
        let currentLevel = remainingLevels[0]
        let nextLevels = Array(remainingLevels.dropFirst())
        
        if !groupingLevels.contains(currentLevel) {
            var sorted = items
            for level in remainingLevels {
                sorted = applySortLevel(sorted, level: level)
            }
            return IntermediateResult(items: sorted)
        }
        
        switch currentLevel {
        case .store:
            return groupByStore(items: items, stores: stores, remainingLevels: nextLevels, listCategories: listCategories, listType: listType)
        case .category:
            return groupByCategory(items: items, stores: stores, remainingLevels: nextLevels, listCategories: listCategories, listType: listType)
        case .rsvp:
            return groupByRsvp(items: items, stores: stores, remainingLevels: nextLevels, listCategories: listCategories, listType: listType)
        default:
            return IntermediateResult(items: items)
        }
    }
    
    private static func groupByStore(
        items: [Item],
        stores: [Store],
        remainingLevels: [SortLevel],
        listCategories: [CategoryDef],
        listType: String
    ) -> IntermediateResult {
        let storeMap = Dictionary(uniqueKeysWithValues: stores.map { ($0.id, $0) })
        
        var grouped: [UUID: [Item]] = [:]
        var ungrouped: [Item] = []
        
        for item in items {
            if let storeId = item.storeId, storeMap[storeId] != nil {
                grouped[storeId, default: []].append(item)
            } else {
                ungrouped.append(item)
            }
        }
        
        let sortedStores = stores.sorted { $0.sortOrder < $1.sortOrder }
        
        var groups: [SortGroup] = []
        for store in sortedStores {
            guard let storeItems = grouped[store.id], !storeItems.isEmpty else { continue }
            
            let subResult = applyRemainingLevels(
                items: storeItems,
                remainingLevels: remainingLevels,
                stores: stores,
                listCategories: listCategories,
                listType: listType
            )
            
            var group = SortGroup(
                key: "store-\(store.id.uuidString)",
                label: store.name,
                color: store.color,
                type: .store,
                items: nil,
                subGroups: nil
            )
            
            if let subGroups = subResult.groups {
                var finalSubGroups = subGroups
                if let subUngrouped = subResult.ungrouped, !subUngrouped.isEmpty {
                    finalSubGroups.append(SortGroup(
                        key: "store-\(store.id.uuidString)-other",
                        label: "Other",
                        color: "#9e9e9e",
                        type: .category,
                        items: subUngrouped,
                        subGroups: nil
                    ))
                }
                group.subGroups = finalSubGroups
            } else {
                group.items = subResult.items
            }
            
            groups.append(group)
        }
        
        if !ungrouped.isEmpty {
            let subResult = applyRemainingLevels(
                items: ungrouped,
                remainingLevels: remainingLevels,
                stores: stores,
                listCategories: listCategories,
                listType: listType
            )
            
            var unassignedGroup = SortGroup(
                key: "store-unassigned",
                label: "Unassigned",
                color: nil,
                type: .store,
                items: nil,
                subGroups: nil
            )
            
            if let subGroups = subResult.groups {
                var finalSubGroups = subGroups
                if let subUngrouped = subResult.ungrouped, !subUngrouped.isEmpty {
                    finalSubGroups.append(SortGroup(
                        key: "store-unassigned-other",
                        label: "Other",
                        color: "#9e9e9e",
                        type: .category,
                        items: subUngrouped,
                        subGroups: nil
                    ))
                }
                unassignedGroup.subGroups = finalSubGroups
            } else {
                unassignedGroup.items = subResult.items ?? ungrouped
            }
            
            groups.append(unassignedGroup)
        }
        
        return IntermediateResult(groups: groups, ungrouped: [])
    }
    
    private static func groupByCategory(
        items: [Item],
        stores: [Store],
        remainingLevels: [SortLevel],
        listCategories: [CategoryDef],
        listType: String
    ) -> IntermediateResult {
        let categoryInfo = getCategoryInfo(listCategories: listCategories, listType: listType)
        let validCategories = Set(categoryInfo.orderedKeys)
        
        var grouped: [String: [Item]] = [:]
        var other: [Item] = []
        
        for item in items {
            if let cat = item.category, validCategories.contains(cat) {
                grouped[cat, default: []].append(item)
            } else {
                other.append(item)
            }
        }
        
        var groups: [SortGroup] = []
        for categoryKey in categoryInfo.orderedKeys {
            guard let categoryItems = grouped[categoryKey], !categoryItems.isEmpty else { continue }
            
            let subResult = applyRemainingLevels(
                items: categoryItems,
                remainingLevels: remainingLevels,
                stores: stores,
                listCategories: listCategories,
                listType: listType
            )
            
            var group = SortGroup(
                key: "category-\(categoryKey)",
                label: categoryInfo.labels[categoryKey] ?? categoryKey,
                color: categoryInfo.colors[categoryKey] ?? "#9e9e9e",
                type: .category,
                items: nil,
                subGroups: nil
            )
            
            if let subGroups = subResult.groups {
                group.subGroups = subGroups
            } else {
                group.items = subResult.items
            }
            
            groups.append(group)
        }
        
        var processedOther = other
        if !other.isEmpty && !remainingLevels.isEmpty {
            let subResult = applyRemainingLevels(
                items: other,
                remainingLevels: remainingLevels,
                stores: stores,
                listCategories: listCategories,
                listType: listType
            )
            processedOther = subResult.items ?? other
        }
        
        if !processedOther.isEmpty {
            groups.append(SortGroup(
                key: "category-other",
                label: "Other",
                color: "#9e9e9e",
                type: .category,
                items: processedOther,
                subGroups: nil
            ))
        }
        
        return IntermediateResult(groups: groups, ungrouped: [])
    }
    
    private static func groupByRsvp(
        items: [Item],
        stores: [Store],
        remainingLevels: [SortLevel],
        listCategories: [CategoryDef],
        listType: String
    ) -> IntermediateResult {
        let rsvpOrder: [(status: String, label: String, color: String)] = [
            ("confirmed", "Confirmed", "#4caf50"),
            ("maybe", "Maybe", "#ff9800"),
            ("invited", "Invited", "#42a5f5"),
            ("declined", "Declined", "#f44336"),
            ("not_invited", "Not Yet Invited", "#9e9e9e")
        ]
        
        var grouped: [String: [Item]] = [:]
        
        for item in items {
            let status = item.rsvpStatus ?? "invited"
            grouped[status, default: []].append(item)
        }
        
        var groups: [SortGroup] = []
        for rsvp in rsvpOrder {
            guard let rsvpItems = grouped[rsvp.status], !rsvpItems.isEmpty else { continue }
            
            let subResult = applyRemainingLevels(
                items: rsvpItems,
                remainingLevels: remainingLevels,
                stores: stores,
                listCategories: listCategories,
                listType: listType
            )
            
            var group = SortGroup(
                key: "rsvp-\(rsvp.status)",
                label: rsvp.label,
                color: rsvp.color,
                type: .rsvp,
                items: nil,
                subGroups: nil
            )
            
            if let subGroups = subResult.groups {
                group.subGroups = subGroups
            } else {
                group.items = subResult.items
            }
            
            groups.append(group)
        }
        
        return IntermediateResult(groups: groups, ungrouped: [])
    }
    
    private struct CategoryInfo {
        var labels: [String: String]
        var colors: [String: String]
        var orderedKeys: [String]
    }
    
    private static func getCategoryInfo(listCategories: [CategoryDef], listType: String) -> CategoryInfo {
        let categories = listCategories.isEmpty ? CategoryDefinitions.defaults : listCategories
        
        var labels: [String: String] = [:]
        var colors: [String: String] = [:]
        var orderedKeys: [String] = []
        
        for cat in categories {
            labels[cat.key] = cat.name
            colors[cat.key] = cat.color
            orderedKeys.append(cat.key)
        }
        
        return CategoryInfo(labels: labels, colors: colors, orderedKeys: orderedKeys)
    }
}

File: src/utils/sortPipeline.js
/**
 * Sort pipeline engine for shopping list items.
 * Provides flexible, composable sorting and grouping of items
 * based on a configurable sort pipeline.
 */

import {
  DEFAULT_CATEGORIES,
  getAllCategoryLabels,
  getAllCategoryColors,
  getAllCategoryKeys,
} from './categories.js';
import { getTypeConfig, PACKING_CATEGORIES, TODO_CATEGORIES, PROJECT_CATEGORIES } from './listTypes.js';

/** Valid sort level options */
export const SORT_LEVELS = ['store', 'category', 'name', 'date', 'price', 'rsvp', 'dueDate'];

/** Maximum levels for grouping (Group By section) */
export const MAX_GROUP_LEVELS = 2;

/** Maximum levels for sort-only (Sort By section) */
export const MAX_SORT_ONLY_LEVELS = 2;

/** Maximum total sort levels */
export const MAX_SORT_LEVELS = MAX_GROUP_LEVELS + MAX_SORT_ONLY_LEVELS;

/** System default sort configuration */
export const SYSTEM_DEFAULT_SORT_CONFIG = ['store', 'category', 'name'];

/**
 * Returns the default sort config for a given list type.
 * @param {string} [listType='grocery'] - The list type identifier
 * @returns {string[]} Default sort config for the type
 */
export const getDefaultSortConfig = (listType) => {
  const config = getTypeConfig(listType ?? 'grocery');
  return config.defaultSort;
};

/** Levels that create groups (vs just sorting within existing groups) */
export const GROUPING_LEVELS = ['store', 'category', 'rsvp'];

/**
 * Validates a sort configuration.
 * @param {unknown} config - The configuration to validate
 * @returns {boolean} True if config is a valid sort configuration
 */
export const isValidSortConfig = (config) => {
  if (!Array.isArray(config) || config.length === 0 || config.length > MAX_SORT_LEVELS) {
    return false;
  }
  const seen = new Set();
  for (const level of config) {
    if (typeof level !== 'string' || !SORT_LEVELS.includes(level)) {
      return false;
    }
    if (seen.has(level)) {
      return false;
    }
    seen.add(level);
  }
  return true;
};

/**
 * Normalizes any input into a valid sort configuration.
 * Filters invalid values, removes duplicates, truncates to MAX_SORT_LEVELS,
 * and falls back to system default if empty.
 * @param {unknown} config - The configuration to normalize
 * @returns {string[]} A valid sort configuration array
 */
export const normalizeSortConfig = (config) => {
  if (!Array.isArray(config)) {
    return [...SYSTEM_DEFAULT_SORT_CONFIG];
  }

  const seen = new Set();
  const normalized = [];

  for (const level of config) {
    if (
      typeof level === 'string' &&
      SORT_LEVELS.includes(level) &&
      !seen.has(level)
    ) {
      seen.add(level);
      normalized.push(level);
      if (normalized.length === MAX_SORT_LEVELS) break;
    }
  }

  if (normalized.length === 0) {
    return [...SYSTEM_DEFAULT_SORT_CONFIG];
  }

  // Safety: ensure grouping levels come before sort-only levels
  const groups = normalized.filter(l => GROUPING_LEVELS.includes(l));
  const sorts = normalized.filter(l => !GROUPING_LEVELS.includes(l));
  return [...groups, ...sorts];
};

/**
 * Sorts items alphabetically by name (case-insensitive).
 * @param {Array<{name: string}>} items - Items to sort
 * @returns {Array} Sorted items (new array)
 */
const sortByName = (items) =>
  [...items].sort((a, b) =>
    a.name.localeCompare(b.name, undefined, { sensitivity: 'base' })
  );

/**
 * Sorts items by added_at date (newest first).
 * @param {Array<{added_at: string}>} items - Items to sort
 * @returns {Array} Sorted items (new array)
 */
const sortByDate = (items) =>
  [...items].sort((a, b) => {
    const dateA = a.addedAt ? new Date(a.addedAt).getTime() : 0;
    const dateB = b.addedAt ? new Date(b.addedAt).getTime() : 0;
    return dateB - dateA;
  });

const sortByPrice = (items) =>
  [...items].sort((a, b) => {
    const priceA = a.price ?? null;
    const priceB = b.price ?? null;
    if (priceA === null && priceB === null) return 0;
    if (priceA === null) return 1;
    if (priceB === null) return -1;
    return priceA - priceB;
  });

/**
 * Sorts items by due date (soonest first, nulls at bottom).
 * @param {Array<{dueDate: string|null}>} items - Items to sort
 * @returns {Array} Sorted items (new array)
 */
const sortByDueDate = (items) =>
  [...items].sort((a, b) => {
    if (!a.dueDate && !b.dueDate) return 0;
    if (!a.dueDate) return 1;
    if (!b.dueDate) return -1;
    return new Date(a.dueDate) - new Date(b.dueDate);
  });

/**
 * Applies a sort-only level to items.
 * @param {Array} items - Items to sort
 * @param {string} level - Sort level ('name', 'date', 'price', or 'dueDate')
 * @returns {Array} Sorted items
 */
const applySortLevel = (items, level) => {
  if (level === 'name') return sortByName(items);
  if (level === 'date') return sortByDate(items);
  if (level === 'price') return sortByPrice(items);
  if (level === 'dueDate') return sortByDueDate(items);
  return items;
};

/**
 * Applies remaining sort levels to items within a group.
 * @param {Array} items - Items to sort
 * @param {string[]} remainingLevels - Remaining sort levels
 * @param {Array} stores - Store objects
 * @param {string|null} parentStoreId - Store ID if under a store group
 * @param {string} listType - List type identifier
 * @param {Array|null} listCategories - Pre-resolved categories for the list
 * @returns {Object} Result with items or groups
 */
const applyRemainingLevels = (items, remainingLevels, stores, parentStoreId, listType, listCategories) => {
  if (remainingLevels.length === 0) {
    return { items };
  }

  const [currentLevel, ...nextLevels] = remainingLevels;

  if (!GROUPING_LEVELS.includes(currentLevel)) {
    let sorted = items;
    for (const level of remainingLevels) {
      sorted = applySortLevel(sorted, level);
    }
    return { items: sorted };
  }

  if (currentLevel === 'store') {
    return groupByStore(items, stores, nextLevels, listType, listCategories);
  }

  if (currentLevel === 'category') {
    return groupByCategory(items, stores, nextLevels, parentStoreId, listType, listCategories);
  }

  if (currentLevel === 'rsvp') {
    return groupByRsvp(items, stores, nextLevels, listType, listCategories);
  }

  return { items };
};

/**
 * Groups items by store.
 * @param {Array} items - Items to group
 * @param {Array} stores - Store objects
 * @param {string[]} remainingLevels - Remaining sort levels after store
 * @param {string} listType - List type identifier
 * @param {Array|null} listCategories - Pre-resolved categories for the list
 * @returns {Object} Result with groups and ungrouped
 */
const groupByStore = (items, stores, remainingLevels, listType, listCategories) => {
  const storeMap = new Map();
  for (const store of stores) {
    storeMap.set(store.id, store);
  }

  const grouped = new Map();
  const ungrouped = [];

  for (const item of items) {
    if (item.store && storeMap.has(item.store)) {
      if (!grouped.has(item.store)) {
        grouped.set(item.store, []);
      }
      grouped.get(item.store).push(item);
    } else {
      ungrouped.push(item);
    }
  }

  const sortedStores = [...stores].sort(
    (a, b) => (a.sort_order ?? 0) - (b.sort_order ?? 0)
  );

  const groups = [];
  for (const store of sortedStores) {
    const storeItems = grouped.get(store.id);
    if (!storeItems || storeItems.length === 0) continue;

    const subResult = applyRemainingLevels(
      storeItems,
      remainingLevels,
      stores,
      store.id,
      listType,
      listCategories
    );

    const group = {
      key: `store-${store.id}`,
      label: store.name,
      color: store.color,
      type: 'store',
    };

    if (subResult.groups) {
      group.subGroups = subResult.groups;
      if (subResult.ungrouped?.length > 0) {
        group.subGroups.push({
          key: `store-${store.id}-other`,
          label: 'Other',
          color: '#9e9e9e',
          type: 'category',
          items: subResult.ungrouped,
        });
      }
    } else {
      group.items = subResult.items;
    }

    groups.push(group);
  }

  let processedUngrouped = ungrouped;
  if (ungrouped.length > 0 && remainingLevels.length > 0) {
    const subResult = applyRemainingLevels(
      ungrouped,
      remainingLevels,
      stores,
      null,
      listType,
      listCategories
    );
    processedUngrouped = subResult.items ?? ungrouped;
  }

  return { groups, ungrouped: processedUngrouped };
};

/**
 * Gets category info for grouping.
 * @param {Array|null} listCategories - Pre-resolved categories for the list
 * @param {string} listType - List type identifier
 * @returns {Object} Category labels, colors, and ordered keys
 */
const getCategoryInfo = (listCategories, listType) => {
  if (listType === 'packing') {
    return {
      labels: Object.fromEntries(PACKING_CATEGORIES.map((c) => [c.key, c.name])),
      colors: Object.fromEntries(PACKING_CATEGORIES.map((c) => [c.key, c.color])),
      orderedKeys: PACKING_CATEGORIES.map((c) => c.key),
    };
  }
  if (listType === 'todo') {
    return {
      labels: Object.fromEntries(TODO_CATEGORIES.map((c) => [c.key, c.name])),
      colors: Object.fromEntries(TODO_CATEGORIES.map((c) => [c.key, c.color])),
      orderedKeys: TODO_CATEGORIES.map((c) => c.key),
    };
  }
  if (listType === 'project') {
    return {
      labels: Object.fromEntries(PROJECT_CATEGORIES.map((c) => [c.key, c.name])),
      colors: Object.fromEntries(PROJECT_CATEGORIES.map((c) => [c.key, c.color])),
      orderedKeys: PROJECT_CATEGORIES.map((c) => c.key),
    };
  }

  const categories = listCategories?.length > 0 ? listCategories : DEFAULT_CATEGORIES;

  return {
    labels: getAllCategoryLabels(categories),
    colors: getAllCategoryColors(categories),
    orderedKeys: getAllCategoryKeys(categories),
  };
};

/**
 * Groups items by category.
 * @param {Array} items - Items to group
 * @param {Array} stores - Store objects
 * @param {string[]} remainingLevels - Remaining sort levels after category
 * @param {string|null} parentStoreId - Store ID if under a store group
 * @param {string} listType - List type identifier
 * @param {Array|null} listCategories - Pre-resolved categories for the list
 * @returns {Object} Result with groups and ungrouped (items with no/unknown category)
 */
const groupByCategory = (items, stores, remainingLevels, parentStoreId, listType, listCategories) => {
  const { labels, colors, orderedKeys } = getCategoryInfo(listCategories, listType);
  const validCategories = new Set(orderedKeys);

  const grouped = new Map();
  const other = [];

  for (const item of items) {
    if (item.category && validCategories.has(item.category)) {
      if (!grouped.has(item.category)) {
        grouped.set(item.category, []);
      }
      grouped.get(item.category).push(item);
    } else {
      other.push(item);
    }
  }

  const groups = [];
  for (const categoryKey of orderedKeys) {
    const categoryItems = grouped.get(categoryKey);
    if (!categoryItems || categoryItems.length === 0) continue;

    const subResult = applyRemainingLevels(
      categoryItems,
      remainingLevels,
      stores,
      parentStoreId,
      listType,
      listCategories
    );

    const group = {
      key: `category-${categoryKey}`,
      label: labels[categoryKey] || categoryKey,
      color: colors[categoryKey] || '#9e9e9e',
      type: 'category',
    };

    if (subResult.groups) {
      group.subGroups = subResult.groups;
    } else {
      group.items = subResult.items;
    }

    groups.push(group);
  }

  let processedOther = other;
  if (other.length > 0 && remainingLevels.length > 0) {
    const subResult = applyRemainingLevels(
      other,
      remainingLevels,
      stores,
      parentStoreId,
      listType,
      listCategories
    );
    processedOther = subResult.items ?? other;
  }

  if (processedOther.length > 0) {
    groups.push({
      key: 'category-other',
      label: 'Other',
      color: '#9e9e9e',
      type: 'category',
      items: processedOther,
    });
  }

  return { groups, ungrouped: [] };
};

/** RSVP status order and display configuration */
const RSVP_ORDER = [
  { status: 'confirmed', label: 'Confirmed', color: '#4caf50' },
  { status: 'maybe', label: 'Maybe', color: '#ff9800' },
  { status: 'invited', label: 'Invited', color: '#42a5f5' },
  { status: 'declined', label: 'Declined', color: '#f44336' },
  { status: 'not_invited', label: 'Not Yet Invited', color: '#9e9e9e' },
];

/**
 * Groups items by RSVP status.
 * @param {Array} items - Items to group
 * @param {Array} stores - Store objects
 * @param {string[]} remainingLevels - Remaining sort levels after rsvp
 * @param {string} listType - List type identifier
 * @param {Array|null} listCategories - Pre-resolved categories for the list
 * @returns {Object} Result with groups
 */
const groupByRsvp = (items, stores, remainingLevels, listType, listCategories) => {
  const grouped = new Map();

  for (const item of items) {
    const status = item.rsvpStatus ?? 'invited';
    if (!grouped.has(status)) {
      grouped.set(status, []);
    }
    grouped.get(status).push(item);
  }

  const groups = [];
  for (const rsvp of RSVP_ORDER) {
    const rsvpItems = grouped.get(rsvp.status);
    if (!rsvpItems || rsvpItems.length === 0) continue;

    const subResult = applyRemainingLevels(
      rsvpItems,
      remainingLevels,
      stores,
      null,
      listType,
      listCategories
    );

    const group = {
      key: `rsvp-${rsvp.status}`,
      label: rsvp.label,
      color: rsvp.color,
      type: 'rsvp',
    };

    if (subResult.groups) {
      group.subGroups = subResult.groups;
    } else {
      group.items = subResult.items;
    }

    groups.push(group);
  }

  return { groups, ungrouped: [] };
};

/**
 * Partition a sort config into group levels and sort-only levels.
 * Maintains the relative order within each partition.
 * @param {string[]} config - Sort configuration array
 * @returns {{groupLevels: string[], sortOnlyLevels: string[]}} Partitioned levels
 */
export function partitionSortConfig(config) {
  const groupLevels = config.filter(l => GROUPING_LEVELS.includes(l));
  const sortOnlyLevels = config.filter(l => !GROUPING_LEVELS.includes(l));
  return { groupLevels, sortOnlyLevels };
}

/**
 * Combine group levels and sort-only levels into a single config array.
 * Groups always come before sort-only levels.
 * @param {string[]} groupLevels - Grouping levels
 * @param {string[]} sortOnlyLevels - Sort-only levels
 * @returns {string[]} Combined sort configuration
 */
export function combineSortConfig(groupLevels, sortOnlyLevels) {
  return [...groupLevels, ...sortOnlyLevels];
}

/**
 * Applies a sort pipeline to items, returning a nested group structure.
 *
 * @param {Array} items - Array of item objects with at least: { id, name, category, store, added_at, isChecked }
 * @param {Array} sortConfig - Array of 1-4 sort level strings (e.g. ['store', 'category', 'name', 'date'])
 * @param {Array} stores - Array of store objects with: { id, name, color, sort_order }
 * @param {string} listType - List type identifier (e.g. 'grocery', 'packing', 'todo')
 * @param {Array|null} listCategories - Pre-resolved categories for the list
 * @returns {Object} Nested structure with groups, subGroups, and items
 *
 * @example
 * // With grouping levels:
 * {
 *   groups: [
 *     { key: 'store-123', label: 'Walmart', type: 'store', subGroups: [...] }
 *   ],
 *   ungrouped: [...]
 * }
 *
 * @example
 * // With only sort levels:
 * {
 *   groups: [],
 *   items: [...]
 * }
 */
export const applySortPipeline = (items, sortConfig, stores = [], listType = 'grocery', listCategories = null) => {
  const config = normalizeSortConfig(sortConfig);
  const safeStores = Array.isArray(stores) ? stores : [];

  if (!items || items.length === 0) {
    return { groups: [], items: [], ungrouped: [] };
  }

  const hasGroupingLevel = config.some((level) =>
    GROUPING_LEVELS.includes(level)
  );

  if (!hasGroupingLevel) {
    let sorted = [...items];
    for (const level of config) {
      sorted = applySortLevel(sorted, level);
    }
    return { groups: [], items: sorted };
  }

  const result = applyRemainingLevels(items, config, safeStores, null, listType, listCategories);

  return {
    groups: result.groups ?? [],
    ungrouped: result.ungrouped ?? [],
  };
};
